### DISCUSSION TITLE
`vim.fs.normalize` interactions with file path URIs

### DISCUSSION THREAD

**Discussion by jamestrew** [neovim/neovim#28139] [Q&A]
Currently `vim.fs.normalize` will mangle file path type URIs. eg: `vim.fs.normalize("oil:///tmp/testing/")` returns `oil:/tmp/testing`.
This behavior falls in line with Python's `os.path.normpath` so in a sense, I can see this being expected behavior.

On the flip side, doing something like `vim.fs.normalize(vim.api.nvim_buf_get_name(0))` is not always a safe operation due to this behavior. Safe as in not getting an invalid path/URI as the return value.

Wanted to get some ideas on the best way to handle these file path URI edge cases. Would it make sense for `vim.fs.normalize` to always just bypass handling URIs? Or is it on users and plugin authors to avoid blindly using `vim.fs.normalize` with the many ways one can get buffer names in vim?

**Comment by dundargoc** [comment]
@justinmk may have some insights or ideas.

**Comment by justinmk** [comment]
> Currently `vim.fs.normalize` will mangle file path type URIs. eg: `vim.fs.normalize("oil:///tmp/testing/")` returns `oil:/tmp/testing`. This behavior falls in line with Python's `os.path.normpath` so in a sense, I can see this being expected behavior.

Yes. "normalize" should be deterministic. For URIs we may need special handling for the "scheme" part. Although AFAIR, `foo:/// => foo:/` is correct (they are equivalent)? But not `foo: => foo:/`.

> On the flip side, doing something like `vim.fs.normalize(vim.api.nvim_buf_get_name(0))` is not always a safe operation due to this behavior. Safe as in not getting an invalid path/URI as the return value.

Example? Note that a "buffer name" is not necessarily a filepath, so it's invalid to call normalize() on a non-filepath buffer name.


> Would it make sense for `vim.fs.normalize` to always just bypass handling URIs?

Perhaps it should only handle `file:` URIs. And for URIs in general we may need `vim.uri.normalize()`. https://github.com/neovim/neovim/pull/24527

**Reply by jamestrew** [reply]
This question stems from this [issue](https://github.com/nvim-telescope/telescope.nvim/pull/3029) in telescope.nvim where we're taking the `file` fields in the list returned from `getmarklist()` and normalizing them with `vim.fs.normalize`.
We normalize these filenames since we previously had issue passing non-expanded paths (`~`) into the `filename` field in the `{list}` param for `setqlist`.

But the current implementation of `vim.fs.normalize` is incompatible with our current approach for marks on a buffer where the buffer name is a file type URI.

> Perhaps it should only handle file: URIs.

Yes, I personally think it would make sense for `vim.fs.normalize` to handle filetype URIs. But I say "filetype URIs" and not `file:` since I believe filetype URIs are not restricted to have a `file` scheme. "Common" examples I've seen recently in the vim/neovim context include `oil:///` and `fugitive:///`.

**Comment by justinmk** [comment]
Improvements to normalize(): https://github.com/neovim/neovim/pull/28203
